edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/planner/HealthRestartPlanner.cs

```diff
@@ -265,22 +274,47 @@ IEnumerable<Task<ICommand>> ApplyRestartPolicy(IEnumerable<IRuntimeModule> modul
             return restart;
         }
 
-        async Task<IEnumerable<ICommand>> ProcessAddedUpdatedModules(
+        async Task<(Option<IEnumerable<ICommand>> upfrontPullCommands, IEnumerable<ICommand> commands)> ProcessAddedUpdatedModules(
             IList<IModule> modules,
             IImmutableDictionary<string, IModuleIdentity> moduleIdentities,
+            IRuntimeInfo runtimeInfo,
             Func<IModuleWithIdentity, Task<ICommand>> createUpdateCommandMaker)
         {
-            // new modules become a command group containing:
-            //   create followed by a start command if the desired
-            //   status is "running"
+            var upfrontPullCommands = new List<Task<ICommand>>();
+
+            // New modules need an image pull, create, and start command. Depending
+            // on `ModuleStartMode`, we will optionally split the image pulls into
+            // a separate collection so that all image pulls can be executed first.
             var addedTasks = new List<Task<ICommand[]>>();
             foreach (IModule module in modules)
             {
                 if (moduleIdentities.TryGetValue(module.Name, out IModuleIdentity moduleIdentity))
                 {
                     var tasks = new List<Task<ICommand>>();
                     var moduleWithIdentity = new ModuleWithIdentity(module, moduleIdentity);
-                    tasks.Add(createUpdateCommandMaker(moduleWithIdentity));
+
+                    Task<ICommand> prepareUpdate = this.commandFactory.PrepareUpdateAsync(module, runtimeInfo);
+                    Task<ICommand> createUpdate = createUpdateCommandMaker(moduleWithIdentity);
+
+                    // TODO ANDREW: Convert to match?
+                    if (this.moduleUpdateMode == ModuleUpdateMode.NonBlocking)
+                    {
+                        IList<Task<ICommand>> cmds = new List<Task<ICommand>> { prepareUpdate, createUpdate };
+                        // Command needs to be grouped so that image pull is
+                        // guaranteed to succeed before we issue a create command.
+                        // This prevents multiple create commands from getting
+                        // executed within aziot-edged if EdgeAgent timesout
+                        // create request (which itself attempts to pull the image)
+                        // and reissues.
+                        Task<ICommand> createOrUpdateWithUpfrontPull = this.commandFactory.WrapAsync(new GroupCommand(await Task.WhenAll(cmds)));
+                        tasks.Add(createOrUpdateWithUpfrontPull);
+                    }
+                    else
+                    {
+                        upfrontPullCommands.Add(this.commandFactory.PrepareUpdateAsync(module, runtimeInfo));
+                        tasks.Add(createUpdateCommandMaker(moduleWithIdentity));
+                    }
```

Comment:
@ancaantochi As we discussed, we were considering utilizing the command factory (which has knowledge of feature flag) to simplify this logic. I was thinking the opposite when I made it like this. For example, before on Create / Update commands we had inner group commands which made extending logic for this feature harder.  It seemed like a better separation of concerns to have the command factory map to actual aziot-edged api calls. And to then have the planner arrange these as needed.

But to think about the case where the command factory does the job of knowing about the feature flag, here is how it may work:
1. Feature flag is plumbed into command factory as it is now already.
2. When Create/Update is called, it will return  2 commands if the feature flag is enabled (one for pull + one for create/update). If feature flag is off then it will return a group command with (pull + create/update).
3. Above requires changes to the interface for command factory to return multiple commands. Planner will still need to order the generated plan correctly.

Based on my idea above, I feel like it is not the way to go. But lmk if you thought of something better.